evcouplings/complex/similarity.py

```diff
@@ -0,0 +1,232 @@
+        sequences above this identity to the query are not considered paralogs
+
+    Returns
+    -------
+    pd.DataFrame
+        with columns created by merging annotation_data and identity_data
+        entries are paralogs found in the same genome as the query id
+    """
+
+    base_query = query_id.split("/")[0]
+
+    # merge the annotation and identity dataframes
+    data = annotation_data.merge(identity_data, on="id")
+
+    # get all the rows that have an id that contains the
+    # query id. This includes the focus sequence and its hit to
+    # itself in the database.
+    query_hits = data[
+        [base_query in str(x) for x in list(data.id)]
+    ]
+
+    # get the species annotation for the query sequence
+    query_species = list(query_hits.species.dropna())
+
+    # get all rows that are from the query species
+    paralogs = data.query("species == @query_species")
+
+    # confirm that paralogs are below the similarity threshold
+    # ie, are diverged in sequence space from the query
+    paralogs = paralogs[paralogs.identity_to_query < identity_threshold]
+    return paralogs
+
+
+def filter_best_reciprocal(alignment, paralogs, most_similar_in_species, allowed_error=0.02):
+
+    """
+    Takes in a dictionary of the best hit to each genome
+    Removes sequences that are not the best reciprocal hit to the query sequence
+
+    Parameters
+    ----------
+    alignment: str
+        path to sequence alignment file
+    paralogs: pd.DataFrame
+        rows correspond to paralogs to the query sequence
+        Created by find_paralogs() function
+    most_similar_in_species: pd.DataFrame
+        contains the id, species name, and percent identity to query
+        for each sequence that was the best hit to the query in its respective species
+    allowed_error: float
+        in order for a sequence to be filtered out of the alignment, it must be more identitical to a
+        paralog sequence than the target sequence by at least this amount
+
+    Returns
+    -------
+      pd.DataFrame
+        contains the id, species name, and percenty identity to query
+        for each sequence that was the best reciprocal hit to the query sequence
+    """
+    ali = Alignment.from_file(open(alignment))
+
+    # Create an n_paralogs x n_sequences ndarray
+    # where entry i,j is percent identity of paralog i to sequence j
+    # note the identity here will be different than for the unfiltered alignment
+
+    # initialize the matrix
+    identity_mat = np.zeros((len(paralogs), len(ali.ids)), dtype=float)
+
+    for idx, paralog_id in enumerate(paralogs.id):
+        # calculate the % identity of every seq in the alignment to current paralog
+        identities = ali.identities_to(ali[ali.id_to_index[paralog_id]])
+
+        # save the results
+        identity_mat[idx, :] = identities
+
+    indices_to_keep = []
+    # for every sequence in the alignment that is the most similar to the query
+    # in its respective species
```

Comment:
I tend to put "..." at the end of the comment to indicate that it is meant to be read in a way that continues inside the loop